### PR TITLE
Colorscheme1.1 (2D segmentation)

### PR DESCRIPTION
**Pull request by lanzhenw**
Support 2D segmentations in colorscheme:

color by field: 

https://github.com/voxel51/fiftyone/assets/17770824/578d1935-b14a-4748-8c97-e7e773d1bdce

color by value: 
use field `maskTargetsColors` if available, then check dataset `defaultMaskTargetsColors`, otherwise use default assigned color based on its intTarget value

https://github.com/voxel51/fiftyone/assets/17770824/4204661f-bc45-423e-8fec-9ac2d972eeaf

color by instance:

https://github.com/voxel51/fiftyone/assets/17770824/1c341cb6-5cb4-49dc-8754-3c9850dc1b95

Save as default + clear default:

https://github.com/voxel51/fiftyone/assets/17770824/1f9533b0-f5b5-486f-a1bb-7eb4d8d663cf

Reset:

https://github.com/voxel51/fiftyone/assets/17770824/791f30bf-e6b2-4463-9a49-14766bae5618

sample colorscheme json with segmentation settings: 
```
   fields: [
    {
      path: "segmentation1",
      fieldColor: "#0000FF",
      targetMasksColors: [
        {
          intTarget: 15,
          color: "#56b4e9",
        },
      ],
    },
    {
      path: "segmentation2",
      fieldColor: "#dddddd",
      targetMasksColors: [
        {
          intTarget: 15,
          color: "#cc79a7",
        },
      ],
    },
  ],
  defaultMaskTargetsColors: [
    {
      intTarget: 11,
      color: "#008c23",
    },
    {
      intTarget: 15,
      color: "#009e74",
    },
  ],
};
```

1. color by field mode: 
-    use the `fieldcolor` of corresponding field to render segmentations
2. color by value mode: 
-    first check if the pixel value is included in the `targetMasksColors` of the field
-    second check if the pixel value is included in the `defaultTargetMasksColors`
3. color by instance mode: 
- use random color to random each pixel

## What changes are proposed in this pull request?
Allow users to customize 2D segmentation colors

## How is this patch tested? If it is not, please explain why.
Create a dataset with semantic segmentations: 
```
import fiftyone.zoo as foz
dataset = foz.load_zoo_dataset("quickstart")
model = foz.load_zoo_model("fcn-resnet50-coco-torch")
dataset.apply_model(model, label_field="segmentations")
```
Create a dataset with video segmentations: 
```
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart-video", max_samples=1)
model = foz.load_zoo_model("fcn-resnet50-coco-torch")
dataset.apply_model(model, label_field="segmentations")
```
You can test this feature by going to the color panel: 
1) Under color by field mode, modify the field color of segmentation fields
2) Under color by value mode, modify the default mask targets colors settings in the global settings page
3) Under color by value mode, go to segmentation field and set mask target colors
4) Toggle between field tab and JSON viewer tab on color panel, the UI should always reflect the changes
5) Save as default and clear should be able to clear `dataset.app_config.colorscheme` settings
6) Clear should be able to reset the current colorscheme to the `dataset.app_config.colorscheme` (if available), or to `fo.config` or fiftyone app default setting.
7) Can launch the app with color settings. session will sync with python SDK through `session.color_scheme`
```
import fiftyone as fo
import fiftyone.zoo as foz
dataset = foz.load_zoo_dataset("quickstart")
dataset.app_config.color_scheme = fo.ColorScheme(
   color_pool=["#8e9aaf", "#cbc0d3", "#efd3d7", "#feeafa", "#dee2ff"],
   color_by="value",
   fields=[
     {
             "path": "segmentation",
             "fieldColor": "#ff00ff",
             "colorByAttribute": "label",
             "valueColors": [],
             "maskTargetsColors": [
               {
                   "intTarget": 13,
                   "color": "#00ff00",
               },
               {
                   "intTarget": 15,
                   "color": "#ff0066",
               }
             ]
         }
     ],
   label_tags=None,
   default_mask_targets_colors = [
   {
       "intTarget": 13,
       "color": "#33ccff"
    }
   ]
)
dataset.save()
session = fo.launch_app(dataset)
```

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [x] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

File: app/packages/core/src/components/ColorModal/GlobalSetting.tsx [hunk 1]
```diff
@@ -10,11 +11,13 @@ import {
   SectionWrapper,
 } from "./ShareStyledDiv";
 import ColorPalette from "./colorPalette/ColorPalette";
+import DefaultMaskTargets from "./colorPalette/DefaultMaskTargets";
 import ShuffleColor from "./controls/RefreshColor";
-import React from "react";
 
 const GlobalSetting = () => {
   const [colorScheme, setColorScheme] = useRecoilState(fos.colorScheme);
+  const [state, setState] = useState({ useColorscaleName: "name" });
```
Comment (lanzhenw) [inline app/packages/core/src/components/ColorModal/GlobalSetting.tsx:19]: need to remove

File: app/packages/core/src/components/ColorModal/colorPalette/FieldsMaskTarget.tsx [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+import { activeColorPath } from "../state";
+import { getRandomColorFromPool, validateIntMask } from "../utils";
+
+type MaskTargetInput = {
+  idx: number;
+  color: string;
+};
+
+const FieldsMaskTargets: React.FC = () => {
+  const isRGBMask = false; // TODO: add condition check;
```
Comment (lanzhenw) [inline app/packages/core/src/components/ColorModal/colorPalette/FieldsMaskTarget.tsx:18]: condition check here if targets is rgbmask

File: app/packages/looker/src/worker/painter.ts [hunk 3]
```diff
@@ -328,6 +320,26 @@ export const PainterFactory = (requestColor) => ({
...
+                  (x) => x.intTarget === targets[i]
+                );
+              }
+
+              // If a customized color setting is found, get the 32-bit color representation.
+              if (colorInfo) {
+                color = get32BitColor(colorInfo.color);
+              }
+            }
+
```
Comment (lanzhenw) [inline app/packages/looker/src/worker/painter.ts:342]: Implementation of color logic in segmentations in worker

File: app/packages/core/src/components/ColorModal/colorPalette/FieldsColorscale.tsx
Comment (lanzhenw) [inline app/packages/core/src/components/ColorModal/colorPalette/FieldsColorscale.tsx:1]: remove it.

File: app/packages/core/src/components/ColorModal/controls/IdxColorList.tsx [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+  useEffect(() => {
+    setInput(initialValue);
+  }, [values]);
+
+  fos.useOutsideClick(wrapperRef, () => {
+    setShowPicker(Array(values?.length ?? 0).fill(false));
+  });
+
+  console.info(input, values);
```
Comment (lanzhenw) [inline app/packages/core/src/components/ColorModal/controls/IdxColorList.tsx:159]: chop

File: app/packages/core/src/components/ColorModal/utils.ts [hunk 5]
```diff
@@ -97,5 +130,109 @@ export const getDisplayName = (path: ACTIVE_FIELD | { path: string }) => {
...
   pool[Math.floor(Math.random() * pool.length)];
+
+export const validateIntMask = (value: number) => {
+  if (!value || value > 255 || value < 1 || !Number.isInteger(value)) {
+    return false;
+  }
+  return true;
+};
+
+const namedColorScales = [
```
Comment (lanzhenw) [inline app/packages/core/src/components/ColorModal/utils.ts:143]: chop

File: app/packages/looker/src/worker/painter.ts [hunk 6]
```diff
@@ -328,6 +331,18 @@ export const PainterFactory = (requestColor) => ({
...
+                defaultSetting?.[targets[i].toString()];
+
+              // If a customized color setting is found, get the 32-bit color representation.
+              if (customColor) {
+                color = get32BitColor(convertToHex(customColor));
+              }
+            }
             overlay[i] = color ? color : getColor(targets[i]);
           }
         }
```
Comment (lanzhenw) [inline app/packages/looker/src/worker/painter.ts:348]: @sashankaryal @benjaminpkane @manivoxel51 
let me know if you think this for loop can be further optimized for performance.

File: app/packages/core/src/components/ColorModal/utils.ts [hunk 7]
```diff
@@ -97,5 +133,22 @@ export const getDisplayName = (path: ACTIVE_FIELD | { path: string }) => {
   return path;
 };
 
-export const getRandomColorFromPool = (pool: readonly string[]) =>
+export const getRandomColorFromPool = (pool: readonly string[]): string =>
   pool[Math.floor(Math.random() * pool.length)];
+
+export const validateIntMask = (value: number) => {
+  if (!value || value > 255 || value < 1 || !Number.isInteger(value)) {
```
Comment (manivoxel51) [inline app/packages/core/src/components/ColorModal/utils.ts:140]:
```suggestion
  if (!value || !Number.isInteger(value) || value > 255 || value < 1) {
```

**Review by manivoxel51 — COMMENTED** [review]
Nice work so far! 💯 

Left some comments - mostly nitpickin but valuable for making the code more maintainable. lmk if you wanna pair up for any of the comments

File: app/packages/core/src/components/ColorModal/FieldSetting.tsx [hunk 8]
```diff
@@ -244,66 +248,78 @@ const FieldSetting = ({ path }: { path: string }) => {
...
-              value={state.useLabelColors}
-              setValue={(v: boolean) => {
-                setSetting((cur) => {
-                  if (!cur) {
-                    cur = { valueColors: [] };
-                  }
+      {coloring.by == COLOR_BY.VALUE &&
+        isTypeValueSupported &&
+        !isHeatmap &&
+        !isSegmentation && (
```
[the remark this answers is not in the document]
Comment (lanzhenw) [reply app/packages/core/src/components/ColorModal/FieldSetting.tsx:254]: I am implementing this in the colorscale PR for less merge conflict.

File: app/packages/core/src/components/ColorModal/colorPalette/DefaultMaskTargets.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+      ),
+    }),
+    [colorScheme.defaultMaskTargetsColors]
+  );
+
+  return (
+    <div>
+      <Checkbox
+        name={`Use custom color for default mask targets`}
+        value={state.useMaskTargetsColors}
```
[the remark this answers is not in the document]
Comment (lanzhenw) [reply app/packages/core/src/components/ColorModal/colorPalette/DefaultMaskTargets.tsx:66]: This checkbox controls the condition for `IdxColorList` component. It doesn't need the condition.

File: app/packages/core/src/components/ColorModal/colorPalette/FieldsMaskTarget.tsx [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+  const index = useMemo(
+    () => colorScheme.fields?.findIndex((s) => s.path == activePath),
+    [activePath]
+  );
+
+  const shouldShowAddButton = Boolean(
+    setting?.maskTargetsColors && setting.maskTargetsColors.length > 0
+  );
+
+  const useFieldMaskColors = useMemo(
```
[the remark this answers is not in the document]
Comment (lanzhenw) [reply app/packages/core/src/components/ColorModal/colorPalette/FieldsMaskTarget.tsx:43]: It's a duplicate with `shouldShowAddButton`. I am keeping one and using the simple form.

File: app/packages/core/src/components/ColorModal/controls/IdxColorList.tsx [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+
+type MaskColorInput = {
+  intTarget: number;
+  color: string;
+};
+
+type Input = {
+  intTarget?: number;
+  color: string;
+};
```
[the remark this answers is not in the document]
Comment (lanzhenw) [reply app/packages/core/src/components/ColorModal/controls/IdxColorList.tsx:33]: Thanks for catching it. `Input` needs its own type, as it allows intTarget to be undefined. `MaskColorInput` should be imported from relay.

File: app/packages/core/src/components/ColorModal/controls/IdxColorList.tsx [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+  );
+
+  // on changing tabs, sync local state with new session values
+  useEffect(() => {
+    setInput(values ?? []);
+  }, [activePath]);
+
+  useEffect(() => {
+    setInput(initialValue);
+  }, [values]);
```
[the remark this answers is not in the document]
Comment (lanzhenw) [reply app/packages/core/src/components/ColorModal/controls/IdxColorList.tsx:161]: i can combine them. when tab changes or when there are changes from session sdk.

File: app/packages/looker/src/worker/painter.ts [hunk 13]
```diff
@@ -328,6 +331,18 @@ export const PainterFactory = (requestColor) => ({
           ) {
             targets[i] = 0;
           } else {
+            if (coloring.by == COLOR_BY.VALUE) {
+              // Attempt to find a color in the fields mask target color settings
+              // If not found, attempt to find a color in the default mask target colors.
+              const customColor =
+                fieldSetting?.[targets[i].toString()] ??
+                defaultSetting?.[targets[i].toString()];
```
Comment (manivoxel51) [inline app/packages/looker/src/worker/painter.ts:344]:
```suggestion
              const target = targets[i].toString()
              const customColor =
                fieldSetting?.[target] || defaultSetting?.[target];
```

**Review by manivoxel51 — APPROVED** [review]
Nice work Lanny! 🍨 

One usability suggestion 

The currently selected tab on the left side of the color modal can get buried under a long scroll, specially if the schema has lot of fields. In order to help the user know "what they are coloring" on the right side of the modal, I suggest adding a title to the top of the modal showing the currenly selected tab text on the left. something like this

Screen Shot 2023-11-15 at 6 25 30 PM

**Review by benjaminpkane — APPROVED** [review]
LGTM!

File: app/packages/looker/src/worker/painter.ts (hunk 6 above)
Comment (benjaminpkane) [reply app/packages/looker/src/worker/painter.ts:356]: Will follow up on this after Thanksgiving! Maybe we can improve a little

**Pull request merged**